File: sdk/csharp/Orbit.Sdk/InstalledStorage.cs
using System.Security.Cryptography;
using System.Text.Json;

namespace Orbit.Sdk;

internal interface IInstalledFiles : IDisposable
{
    string Provider
    {
        get;
    }
    bool Created
    {
        get;
    }
    byte[]? Read();
    void Write(byte[] data);
    void Check();
}

internal sealed class InstalledStorage : ICredentialStorage, IDisposable
{
    private readonly object gate = new();
    private readonly IInstalledFiles files;
    private bool poisoned, disposed;
    internal InstalledRecord Record
    {
        get; private set;
    }
    public StorageCapability Capability => files.Provider == "private_file" ? StorageCapability.PrivateFile : StorageCapability.OperatingSystemProtected;
    internal InstalledStorage(IInstalledFiles files, InstalledScope scope, string? fingerprint, string? provider)
    {
        this.files = files;
        var bytes = files.Read();
        if (bytes == null)
        {
            if (!files.Created)
                throw Storage();
            Record = new(InstalledCodec.Sdk, 2, files.Provider, scope, new(Device.NewInstallation().InstallationId, fingerprint, provider), 0, null, null, null);
            Write(Record);
        }
        else
        {
            try
            {
                Record = InstalledCodec.Decode(bytes, scope, files.Provider, fingerprint, provider);
            }
            finally { CryptographicOperations.ZeroMemory(bytes); }
        }
    }
    private static OrbitException Storage() => new(OrbitError.Storage);
    private void Check()
    {
        if (disposed || poisoned)
            throw Storage();
        files.Check();
    }
    private void Write(InstalledRecord value)
    {
        Check();
        var bytes = InstalledCodec.Encode(value);
        try
        {
            files.Write(bytes);
            Record = value;
        }
        catch (Exception) { poisoned = true; throw Storage(); }
        finally { CryptographicOperations.ZeroMemory(bytes); }
    }
    public long Version
    {
        get
        {
            lock (gate)
            {
                Check();
                return Record.Generation;
            }
        }
    }
    public (long Version, StoredCredential? Credential) Load()
    {
        lock (gate)
        {
            Check();
            return (Record.Generation, Record.Stored);
        }
    }
    public void Save(long version, StoredCredential credential)
    {
        lock (gate)
        {
            Match(version);
            Write(Record with
            {
                Credential = Saved(credential),
                Access = null
            });
        }
    }
    private static InstalledCredential Saved(StoredCredential c) => new(c.ActivationId, c.LicenceId, c.Credential, c.CredentialExpiresAt == 0 ? null : c.CredentialExpiresAt);
    private void Match(long expected)
    {
        Check();
        if (Record.Generation != expected)
            throw new OrbitException(OrbitError.StaleResponse);
    }
    public long Invalidate() => Invalidate(clearPending: true);
    internal long Invalidate(bool clearPending)
    {
        lock (gate)
        {
            if (Record.Generation == long.MaxValue)
                throw Storage();
            Write(Record with
            {
                Generation = Record.Generation + 1,
                Credential = null,
                Access = null,
                PendingActivation = clearPending ? null : Record.PendingActivation
            });
            return Record.Generation;
        }
    }
    internal void DropCache()
    {
        lock (gate)
        {
            Check();
            if (Record.Access != null)
                Write(Record with
                {
                    Access = null
                });
        }
    }
    internal (string OperationId, long Version) Begin(string principal, bool account, string? previous, string? operation)
    {
        lock (gate)
        {
            // Explicit sorted maps provide a deterministic digest without saving principal secrets.
            var data = JsonSerializer.SerializeToUtf8Bytes(new SortedDictionary<string, object?>(StringComparer.Ordinal)
            {
                ["scope"] = new SortedDictionary<string, object?>(StringComparer.Ordinal) { ["api_origin"] = Record.Scope.ApiOrigin, ["issuer"] = Record.Scope.Issuer, ["application_id"] = Record.Scope.ApplicationId, ["environment_id"] = Record.Scope.EnvironmentId },
                ["installation"] = new SortedDictionary<string, object?>(StringComparer.Ordinal) { ["id"] = Record.Installation.Id, ["fingerprint"] = Record.Installation.Fingerprint, ["fingerprint_provider"] = Record.Installation.FingerprintProvider },
                ["principal_kind"] = account ? "account" : "key",
                ["licence_input"] = principal,
                ["previous_credential"] = previous,
                ["credential_mode"] = "persistent"
            });
            var digest = Convert.ToHexStringLower(SHA256.HashData(data));
            CryptographicOperations.ZeroMemory(data);
            var now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            if (now is < 0 or > 253402300799)
                throw new OrbitException(OrbitError.ClockUncertain);
            var pending = Record.PendingActivation;
            if (pending != null)
            {
                if (now < pending.CreatedAt || now - pending.CreatedAt >= 86400)
                    throw new OrbitException(OrbitError.Storage, "pending_activation_expired");
                if (pending.InputDigest != digest || operation != null && operation != pending.OperationId)
                    throw new OrbitException(OrbitError.Storage, "pending_activation");
                operation = pending.OperationId;
            }
            else
            {
                operation ??= Device.NewInstallation().InstallationId;
                pending = new(operation, account ? "account" : "key", digest, now);
            }
            if (Record.Generation == long.MaxValue)
                throw Storage();
            Write(Record with
            {
                Generation = Record.Generation + 1,
                Credential = null,
                Access = null,
                PendingActivation = pending
            });
            return (operation, Record.Generation);
        }
    }
    internal void Commit(long version, StoredCredential credential, JsonElement response, GrantKeys keys, ClockAnchor anchor)
    {
        var token = JsonWire.String(response, "grant");
        var licence = JsonWire.OptionalString(response, "licence_expires_at");
        var access = new InstalledAccess(token, keys.Export(token), licence == null ? null : JsonWire.Timestamp(licence), anchor.ServerSeconds, anchor.WallSeconds, anchor.Now(), DateTimeOffset.UtcNow.ToUnixTimeSeconds());
        lock (gate)
        {
            Match(version);
            if (Record.PendingActivation != null && JsonWire.Field(response, "credential").ValueKind == JsonValueKind.Null)
                throw new OrbitException(OrbitError.Storage, "pending_activation");
            Write(Record with
            {
                Credential = Saved(credential),
                Access = access,
                PendingActivation = null
            });
        }
    }
    internal void Checkpoint(long server, long wall)
    {
        lock (gate)
        {
            Check();
            if (Record.Access is not { } access)
                return;
            if (server < access.ServerHighWater || wall < access.WallHighWater || server > 253402300799 || wall > 253402300799 || Math.Abs((server - access.ReceivedServerTime) - (wall - access.ReceivedWallTime)) > 30)
                throw new OrbitException(OrbitError.ClockUncertain);
            Write(Record with
            {
                Access = access with
                {
                    ServerHighWater = server,
                    WallHighWater = wall
                }
            });
        }
    }
    public void Dispose()
    {
        lock (gate)
        {
            if (disposed)
                return;
            disposed = true;
            Record = Record with
            {
                Credential = null,
                Access = null,
                PendingActivation = null
            };
            files.Dispose();
        }
    }
}

File: sdk/go/installed_storage.go
package orbit

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"math"
	"sync"
	"time"
)

type installedFiles interface {
	read() ([]byte, error)
	write([]byte) error
	check() error
	close() error
}
type installedStorage struct {
	mu       sync.Mutex
	files    installedFiles
	record   installedRecord
	poisoned bool
	closed   bool
}

func (s *installedStorage) checkLocked() error {
	if s.closed || s.poisoned || s.files.check() != nil {
		return ErrStorage
	}
	return nil
}
func (s *installedStorage) writeLocked(record installedRecord) error {
	if err := s.checkLocked(); err != nil {
		return err
	}
	data, err := json.Marshal(record)
	if err != nil || len(data) > installedLimit {
		return ErrStorage
	}
	defer clear(data)
	if s.files.write(data) != nil {
		s.poisoned = true
		return ErrStorage
	}
	s.record = record
	return nil
}
func (s *installedStorage) Load() (uint64, *StoredCredential, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := s.checkLocked(); err != nil {
		return 0, nil, err
	}
	return s.record.Generation, s.record.storedCredential(), nil
}
func (s *installedStorage) Version() (uint64, error) {
	version, _, err := s.Load()
	return version, err
}
func (s *installedStorage) Save(version uint64, credential StoredCredential) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if version != s.record.Generation {
		return ErrStaleResponse
	}
	r := s.record
	r.Credential, r.Access = installedCredentialFrom(&credential), nil
	return s.writeLocked(r)
}
func installedCredentialFrom(c *StoredCredential) *installedCredential {
	if c == nil {
		return nil
	}
	r := &installedCredential{ActivationID: c.ActivationID, LicenceID: c.LicenceID, Bearer: c.Credential}
	if c.CredentialExpiresAt != 0 {
		expiry := c.CredentialExpiresAt
		r.ExpiresAt = &expiry
	}
	return r
}
func (s *installedStorage) Invalidate() (uint64, error) { return s.invalidate(true) }
func (s *installedStorage) invalidate(clearPending bool) (uint64, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	r := s.record
	if r.Generation == math.MaxInt64 {
		return 0, ErrStorage
	}
	r.Generation++
	r.Credential, r.Access = nil, nil
	if clearPending {
		r.Pending = nil
	}
	return r.Generation, s.writeLocked(r)
}
func (s *installedStorage) dropCache() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.record.Access == nil {
		return s.checkLocked()
	}
	r := s.record
	r.Access = nil
	return s.writeLocked(r)
}
func (s *installedStorage) begin(key, licence, previous, operation string) (string, uint64, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	r := s.record
	kind, input := "key", key
	if licence != "" {
		kind, input = "account", licence
	}
	// Marshaling maps recursively sorts keys; retain only this scoped digest.
	encoded, err := json.Marshal(map[string]any{"scope": map[string]any{"api_origin": r.Scope.APIOrigin, "issuer": r.Scope.Issuer, "application_id": r.Scope.ApplicationID, "environment_id": r.Scope.EnvironmentID}, "installation": map[string]any{"id": r.Installation.ID, "fingerprint": r.Installation.Fingerprint, "fingerprint_provider": r.Installation.FingerprintProvider}, "principal_kind": kind, "licence_input": input, "previous_credential": optionalString(previous), "credential_mode": "persistent"})
	if err != nil {
		return "", 0, ErrConfiguration
	}
	digest := sha256.Sum256(encoded)
	clear(encoded)
	hexDigest := hex.EncodeToString(digest[:])
	now := time.Now().Unix()
	if now < 0 {
		return "", 0, ErrClockUncertain
	}
	if p := r.Pending; p != nil {
		if now < p.CreatedAt || now-p.CreatedAt >= 86400 {
			return "", 0, ErrPendingActivationExpired
		}
		if p.InputDigest != hexDigest || p.PrincipalKind != kind || operation != "" && operation != p.OperationID {
			return "", 0, ErrPendingActivation
		}
		operation = p.OperationID
	} else {
		if operation == "" {
			device, err := NewInstallation()
			if err != nil {
				return "", 0, err
			}
			operation = device.InstallationID
		}
		r.Pending = &installedPending{OperationID: operation, PrincipalKind: kind, InputDigest: hexDigest, CreatedAt: now}
	}
	if r.Generation == math.MaxInt64 {
		return "", 0, ErrStorage
	}
	r.Generation++
	r.Credential, r.Access = nil, nil
	return operation, r.Generation, s.writeLocked(r)
}
func (s *installedStorage) commit(version uint64, credential *StoredCredential, response json.RawMessage, keys grantKeys, anchor *timeAnchor) error {
	var reply grantReply
	if decodeJSON(response, &reply) != nil || reply.Grant == nil {
		return ErrStorage
	}
	header, _, err := parseHeader(*reply.Grant)
	if err != nil {
		return ErrStorage
	}
	key := keys[header.KeyID]
	if key == nil {
		return ErrStorage
	}
	set, err := json.Marshal(map[string]any{"keys": []jsonWebKey{{KeyType: "EC", Curve: "P-256", Algorithm: "ES256", Purpose: "sig", KeyID: header.KeyID, X: base64.RawURLEncoding.EncodeToString(key.X.FillBytes(make([]byte, 32))), Y: base64.RawURLEncoding.EncodeToString(key.Y.FillBytes(make([]byte, 32)))}}})
	if err != nil {
		return ErrStorage
	}
	now, err := anchor.now()
	if err != nil {
		return err
	}
	var licenceExpiry *int64
	if reply.LicenceExpiresAt != nil {
		expiry, err := timestamp(*reply.LicenceExpiresAt)
		if err != nil {
			return err
		}
		licenceExpiry = &expiry
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if version != s.record.Generation {
		return ErrStaleResponse
	}
	r := s.record
	r.Credential, r.Pending = installedCredentialFrom(credential), nil
	r.Access = &installedAccess{JWS: *reply.Grant, JWKS: set, LicenceExpiresAt: licenceExpiry, ReceivedServerTime: anchor.server, ReceivedWallTime: anchor.wall, ServerHighWater: now, WallHighWater: time.Now().Unix()}
	return s.writeLocked(r)
}
func (s *installedStorage) checkpoint(server, wall int64) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.record.Access == nil {
		return s.checkLocked()
	}
	r := s.record
	access := *r.Access
	if server < access.ServerHighWater || wall < access.WallHighWater || server > 253402300799 || wall > 253402300799 ||
		(server-access.ReceivedServerTime)-(wall-access.ReceivedWallTime) < -30 || (server-access.ReceivedServerTime)-(wall-access.ReceivedWallTime) > 30 {
		return ErrClockUncertain
	}
	access.ServerHighWater, access.WallHighWater = server, wall
	r.Access = &access
	return s.writeLocked(r)
}
func (s *installedStorage) close() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.closed {
		return nil
	}
	s.closed = true
	s.record = installedRecord{}
	return s.files.close()
}
